decider-bcl2fastq/src/main/java/ca/on/oicr/pde/deciders/utils/BarcodeAndBasesMask.java

```diff
@@ -0,0 +1,178 @@
+package ca.on.oicr.pde.deciders.utils;
+
+import ca.on.oicr.pde.deciders.data.Barcode;
+import ca.on.oicr.pde.deciders.data.BasesMask;
+import ca.on.oicr.pde.deciders.exceptions.DataMismatchException;
+import com.google.common.base.Joiner;
+import com.google.common.collect.Iterables;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ *
+ * @author mlaszloffy
+ */
+public class BarcodeAndBasesMask {
+
+    public static Barcode applyBasesMask(Barcode barcode, BasesMask basesMask) throws DataMismatchException {
+        return applyBasesMask(barcode, basesMask, true);
+    }
+
+    public static Barcode tryApplyingBasesMask(Barcode barcode, BasesMask basesMask) throws DataMismatchException {
+        return applyBasesMask(barcode, basesMask, false);
+    }
+
+    private static Barcode applyBasesMask(Barcode barcode, BasesMask basesMask, boolean strict) throws DataMismatchException {
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this method to reduce its Cognitive Complexity from 35 to the 15 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)